vpr/src/analytical_place/analytical_solver.cpp

```diff
@@ -0,0 +1,264 @@
+            // node.
+            // Using the weight from FastPlace
+            double w = static_cast<double>(num_pins) / static_cast<double>(num_pins - 1);
+            size_t star_node_id = num_moveable_blocks_ + star_node_offset;
+            for (APPinId pin_id : netlist_.net_pins(net_id)) {
+                APBlockId blk_id = netlist_.pin_block(pin_id);
+                add_connection_to_system(star_node_id, blk_id, w);
+            }
+            star_node_offset++;
+        } else {
+            // Create a clique connection where every block in a net connects
+            // exactly once to every other block in the net.
+            // Using the weight from FastPlace
+            double w = 1.0 / static_cast<double>(num_pins - 1);
+            for (size_t ipin_idx = 0; ipin_idx < num_pins; ipin_idx++) {
+                APPinId first_pin_id = netlist_.net_pin(net_id, ipin_idx);
+                APBlockId first_blk_id = netlist_.pin_block(first_pin_id);
+                for (size_t jpin_idx = ipin_idx + 1; jpin_idx < num_pins; jpin_idx++) {
+                    APPinId second_pin_id = netlist_.net_pin(net_id, jpin_idx);
+                    APBlockId second_blk_id = netlist_.pin_block(second_pin_id);
+                    // Make sure that the first node is moveable. This makes
+                    // creating the connection easier.
+                    if (netlist_.block_mobility(first_blk_id) == APBlockMobility::FIXED) {
+                        // If both blocks are fixed, no connection needs to be
+                        // made; just continue.
+                        if (netlist_.block_mobility(second_blk_id) == APBlockMobility::FIXED) {
+                            continue;
+                        }
+                        // If the second block is moveable, swap the first and
+                        // second block so the first block is the moveable one.
+                        std::swap(first_blk_id, second_blk_id);
+                    }
+                    size_t first_row_id = (size_t)blk_id_to_row_id_[first_blk_id];
+                    add_connection_to_system(first_row_id, second_blk_id, w);
+                }
+            }
+        }
+    }
+
+    // Make sure that the number of star nodes created matches the number of
+    // star nodes we pre-calculated we would have.
+    VTR_ASSERT_SAFE(num_star_nodes == star_node_offset);
+
+    // Populate the A_sparse matrix using the triplets.
+    A_sparse.setFromTriplets(tripletList.begin(), tripletList.end());
+}
+
+/**
+ * @brief Helper method to update the linear system with anchors to the current
+ *        partial placement.
+ *
+ * For each moveable block (with row = i) in the netlist:
+ *      A[i][i] = A[i][i] + coeff_pseudo_anchor;
+ *      b[i] = b[i] + pos[block(i)] * coeff_pseudo_anchor;
+ * Where coeff_pseudo_anchor grows with each iteration.
+ *
+ * This is basically a fast way of adding a connection between a moveable block
+ * and a fixed block.
```

Comment:
Great comment!  

Would like to see something like this for the net coefficient code ...

Reply:
Wouldn't this function be shorter (need fewer inputs) if it was part of the analytical solver class?  
Should comment the inputs (names are very good, so short comments probably OK but I still think it's good to have a full doxygen comment).

Reply:
Sadly no. It would only save 2 arguments. It also does not make sense to use this method outside of this file's context.